### PR TITLE
expose template bank parameters in pycbc_coinc_findtrigs

### PR DESCRIPTION
**Pull request by ahnitz**
This makes the template bank parameters (mass1, mass2, spin1z, spin2z) visible within pycbc_coinc_findtrigs, so they can now be used by ranking statistic classes.

### CODE REVIEW

**Comment by tdent** [comment]
Hi Alex .. OK, the intention is clear but I find the ReadByTemplate methods rather cryptic. 
Eg in get_data, what is 'num', what does 'ref' represent, and then what is happening with the double indexing in 'return self.file[ref][ref]' ? 
Can you put in some comments (either formal function definitions or just explanations) for the variables and methods to address questions like this. 
Also syntax mixing %s formatting and string addition is hard to parse, eg 
 key = '%s/search/' % self.ifo + '%s_time' 
 key % 'start' , key % 'end' 
is overkill .. why not 
 key = self.ifo + '/search/' 
 key + 'start_time', key + 'end_time' ?

Also, please also make sure the file is free of whitespace errors (instances of ' \n').

**Comment by ahnitz** [comment]
Hi Tom,

I've added some documentation to these methods.

**Comment by tdent** [comment]
Hi Alex - taking a break from the 'lecture week' in Scotland I *may* be getting somewhere with this.  Start with get_data(), this first does, e.g., 
 self.file['H1/end_time_template'][num] 
I haven't looked at an 'end_time_template' dataset in an hdf file, is this standard and what does it contain?  Looking at a 'HDF_TRIGGER_MERGE' file I get a bunch of things called '<HDF5 region reference>'.  The documentation tells me this is something like a numpy slice.  So then we have the n'th such slice where n=num .. ? 
So, if purely by guesswork I consider that this might slice on the indices of the end_time dataset at which triggers generated by template n (by some hopefully consistent numbering of templates) can be found, then 
 self.file['H1/end_time'][num]
does indeed give me the end times for template n.  
(I can't tell what self.file[ref][ref] would have given me though ...) 

This is OK iff the file is of the exact form you envisage with a column_template dataset (.. attribute? reference?) for every column we want to obtain.  
(.. BTW, is this really preferable to just having one single template_ref reference which could be used for all columns??)  
Thus, I think it would be good to specify that the file has to be of this form right at the beginning of the ReadByTemplate class where the initialization of self.file is done, i.e. have a comment to the __init__ function such as 'file must be a HDF_TRIGGER_MERGE file with a format to be specified at (such and such location)' and eventually the format will be documented there. 

For the 'set_template()' method, the documentation says it returns 'trigger_id' but the return statement is 
return self.keep + self.file['%s/template_boundaries' % self.ifo][num]
which I don't understand as I don't know what the template_boundaries dataset contains.  It would be better to have a line which calculates trigger_id via that formula, with a comment, then return trigger_id.

**Comment by ahnitz** [comment]
On Thu, Jun 11, 2015 at 10:42 AM, tdent <notifications@github.com> wrote:

> Hi Alex - taking a break from the 'lecture week' in Scotland I *may* be
> getting somewhere with this. Start with get_data(), this first does, e.g.,
> self.file['H1/end_time_template'][num]
> I haven't looked at an 'end_time_template' dataset in an hdf file, is this
> standard and what does it contain? Looking at a 'HDF_TRIGGER_MERGE' file I
> get a bunch of things called ''. The documentation tells me this is
> something like a numpy slice. So then we have the n'th such slice where
> n=num .. ?
> So, if purely by guesswork I consider that this might slice on the indices
> of the end_time dataset at which triggers generated by template n (by some
> hopefully consistent numbering of templates) can be found, then
> self.file['H1/end_time'][num]
> does indeed give me the end times for template n.
>
That is correct.

>
> (I can't tell what self.file[ref][ref] would have given me though ...)
>

The hdf references actually refer to both the key and the subset, so the
fil[ref] returns the group, and fil[ref][ref] returns the subset of that
group. That's confusing, so I've removed it.

>  This is OK iff the file is of the exact form you envisage with a
> column_template dataset (.. attribute? reference?) for every column we want
> to obtain.
>
> Yes, it is called a region reference, if you ever have to look it up and
its functionality. Technically, I am storing an array of region references.

> (.. BTW, is this really preferable to just having one single template_ref
> reference which could be used for all columns??)
>
I wish it worked that way, unfortunately it does not, they are tied to the
exact dataset. I think this is for technical reason, as it is not really a
simple slice, given that the dataset can be compressed, byte reordered, etc
and so the optimal read pattern may be different for each column, and
definitely is not guaranteed to the be same. The region reference is
designed for fast (optimized) data access, it is not entirely analogous to
a slice. For a slice, the library has to figure which data from disk it
needs to read in based on your indices, the region reference has that
information precalculated.

>
> Thus, I think it would be good to specify that the file has to be of this
> form right at the beginning of the ReadByTemplate class where the
> initialization of self.file is done, i.e. have a comment to the *init*
> function such as 'file must be a HDF_TRIGGER_MERGE file with a format to be
> specified at (such and such location)' and eventually the format will be
> documented there.
>
I think the goal should be to document the formats in the main
documentation page, and link from there. Otherwise it may get a bit
unwieldy, and this will need to be used it other codes as well. Something
like "requires X format, defined at link here".

>  For the 'set_template()' method, the documentation says it returns
> 'trigger_id' but the return statement is
> return self.keep + self.file['%s/template_boundaries' % self.ifo][num]
> which I don't understand as I don't know what the template_boundaries
> dataset contains. It would be better to have a line which calculates
> trigger_id via that formula, with a comment, then return trigger_id.
>

Ok, done.

**Comment by ahnitz** [comment]
Hey Tom,

Any other changes of the code before I can push this?

**Comment by tdent** [comment]
I don't see a further commit for neatening the return of set_template() and explaining what the template_boundaries are but I'll assume you've got it done somewhere on a branch, so please go ahead once that is included in the merge.

**Comment by ahnitz** [comment]
Oh, sorry I forgot to push that. It is pushed now.
